src/Typechecker.ml

```diff
@@ -5,117 +5,158 @@ open Ast.Expr
 
 type error = string
 
-let rec check_open delta gamma expr typ =
+let rec check_open delta gamma Location.{ data = expr; loc }
+    Location.{ data = typ; loc = t_loc } =
   match expr with
   | Unit ->
       Result.ok_if_true
-        ([%equal: Type.t] typ Type.Unit)
-        ~error:"Expected unit type"
+        ([%equal: Type.t'] typ Type.Unit)
+        ~error:(Location.pp ~msg:"Expected unit type" loc)
   | Pair (e1, e2) -> (
       match typ with
       | Type.Prod (t1, t2) ->
           let%map () = check_open delta gamma e1 t1
           and () = check_open delta gamma e2 t2 in
           ()
-      | _ -> Result.fail "Expected product type" )
+      | _ -> Result.fail @@ Location.pp ~msg:"Expected product type" loc )
   | Fst pe -> (
       let%bind t = infer_open delta gamma pe in
-      match t with
+      match t.Location.data with
```

Comment:
I thought we agreed to not use projections here, change`let%bind t` to pattern-match in-place. This comment applies to the rest of this file.